nvflare/app_common/ccwf/swarm_client_ctl.py

```diff
@@ -0,0 +1,620 @@
+        self.log_info(fl_ctx, f"aggr result: {aggr_result}")
+        global_weights = self.shareable_generator.shareable_to_learnable(aggr_result, fl_ctx)
+        self.record_last_result(fl_ctx, gatherer.for_round, global_weights)
+
+        # are we done with training?
+        num_rounds_done = gatherer.for_round - self.get_config_prop(Constant.START_ROUND, 0) + 1
+        if num_rounds_done >= self.get_config_prop(AppConstants.NUM_ROUNDS):
+            self.log_info(fl_ctx, f"Swarm Learning Done: number of rounds completed {num_rounds_done}")
+
+            # determine the best global result
+            self._distribute_final_results(aggr_result, fl_ctx)
+            return
+
+        # continue next round
+        next_round_data = self.shareable_generator.learnable_to_shareable(global_weights, fl_ctx)
+        assert isinstance(next_round_data, Shareable)
+
+        best_round = aggr_result.get_header(Constant.ROUND)
+        best_metric = aggr_result.get_header(Constant.METRIC)
+        best_client = aggr_result.get_header(Constant.CLIENT)
+
+        if best_client:
+            next_round_data.set_header(Constant.ROUND, best_round)
+            next_round_data.set_header(Constant.CLIENT, best_client)
+            next_round_data.set_header(Constant.METRIC, best_metric)
+
+        self._scatter(next_round_data, gatherer.for_round + 1, gatherer.fl_ctx)
+
+    def _ask_to_share_best_result(self, client: str, metric, fl_ctx: FLContext):
+        # other client has best model - ask it to distribute its result
+        self.log_info(fl_ctx, f"client {client} has the best metric {metric} - ask it to share result")
+        resp = self.engine.send_aux_request(
+            targets=[client],
+            topic=self.topic_for_my_workflow(Constant.TOPIC_SHARE_RESULT),
+            request=Shareable(),
+            timeout=self.final_result_send_timeout,
+            fl_ctx=fl_ctx,
+        )
+
+        assert isinstance(resp, dict)
+        reply = resp.get(client)
+        if not reply:
+            self.log_error(fl_ctx, f"failed to ask client {client} to share final result")
+            return
+
+        if not isinstance(reply, Shareable):
+            self.log_error(fl_ctx, f"client {client} failed to respond to share final result request")
+            return
+
+        rc = reply.get_return_code()
+        if rc != ReturnCode.OK:
+            self.log_error(fl_ctx, f"client {client} failed to respond to share final result request: {rc}")
+
+    def _distribute_final_results(self, aggr_result: Shareable, fl_ctx: FLContext):
+        best_client = aggr_result.get_header(Constant.CLIENT)
+        best_metric = aggr_result.get_header(Constant.METRIC)
+
+        if best_client:
+            if best_client == self.me:
+                # I have the best model
+                self.log_info(fl_ctx, f"I have global best metric {best_metric}")
+                self.broadcast_final_result(
```

Comment:
No, the logic here is:
- Do I have the overall best model? If so, I will broadcast to others
- If not, I will ask the client that has the best model to broadcast to others

Reply:
No, client 2 won't ask client 1 for the best model. There are only two cases:
- Client1 has the best model, and it just broadcasts to other clients
- Client 1 doesn't have the best model but it knows who has it. It sends a message to that client to ask it to broadcast.
No client proactively asks other client for best model. Every client simply waits for the best model to come.

Reply:
As I said, client 1 doesn't have the best model but it knows who has it (say client B has it). It asks Client B to please BROADCAST the model it holds to other clients. All other clients simply WAIT for Client B's broadcast. No one asks for models because they don't even know where to ask. They wait for the model holder to send the model to them.